File: Baekjoon/Python/BOJ_PS/Recursion/Q1629.py
# 곱셈 / 재귀
import sys 
input = sys.stdin.readline

a, b, c = map(int, input().split())

def oprt(a, b, c):
    if b == 1:
        return a 
    else:
        temp = oprt(a, b//2, c)
        if b % 2 == 0:
            return temp * temp % c
        else:
            return temp * temp * a % c

print(oprt(a,b,c))

File: Baekjoon/Swift/BOJ_PS/BOJ_PS/Recursion/Q1629.swift
//
//  Q1629.swift
//  BOJ_PS
//
//  Created by 김태훈 on 2021/05/25.
//  곱셈 / 분할정복, 재귀
/*
 a^b를 구하는 식에서
 b가 짝수일 때, a^(b/2) * a^(b/2)
 b가 홀수일 때, a^(b/2) * a^(b/2) * a
 */
let nums = readLine()!.split(separator: " ").map { Int(String($0))! }

func oprt(_ a: Int, _ b: Int, _ c: Int) -> Int {
    if b == 1 {
        return a % c
    }
    var temp = oprt(a, b/2, c)
    temp = temp * temp % c
    guard b % 2 == 1 else { return temp }
    return temp * a % c
}

print(oprt(nums[0], nums[1], nums[2]))
